app/services/rate_limiter.rb

```diff
@@ -59,6 +59,7 @@ def remaining_count
   end
 
   def expired?
+    return nil if expires_at.nil?
     expires_at <= Time.zone.now
   end
```

Comment:
I feel like this turns a boolean predicate method into a ternary (true/false/nil) with different meanings.

What if we added smarts to the `#reset` method like this that let it no-op gracefully?

```diff
diff --git a/app/services/rate_limiter.rb b/app/services/rate_limiter.rb
index 6e7a0f8b6..0bdb673b0 100644
--- a/app/services/rate_limiter.rb
+++ b/app/services/rate_limiter.rb
@@ -111,9 +111,11 @@ class RateLimiter
     self
   end
 
-  def reset!
-    REDIS_THROTTLE_POOL.with do |client|
-      client.del(key)
+  def reset!(force_reset: false)
+    if expires_at.future? || force_reset
+      REDIS_THROTTLE_POOL.with do |client|
+        client.del(key)
+      end
     end
```

And the `force_reset` would let us clean things up in specs if needed, etc?

Reply:
I think all uses of `reset!` would want to use `force_reset`, so I'm leaning against including that as an optional parameter.

It does change the meaning to have three states, but I think it is closer to the right model since `nil` represents that there currently is no expiration. The bugs uncovered in the existing tests I think nudge me towards returning `nil` as well since it was using the existence of an `expires_at` value to test whether the rate limiter was working.

`OtpRateLimiter` is the only one that uses expiration in this way, so I think it is safe from an implementation perspective too.

Reply:
Ok! I guess I don't have as tight an understanding of this class, but if you think that kind of overload is the most straightforward, let's do it

Reply:
Apologies for merging too soon. I do think it is the simpler path at the moment since I'm not confident how we should think about whether something is "expired" when a value doesn't exist. We currently have a default value [here](https://github.com/18F/identity-idp/blob/bd92be16aa6cb431a199b2365ffd8a2a5622ebe6/app/views/idv/session_errors/failure.html.erb#L19) when showing a countdown, but it might make sense to use the nil state to not show "try again in 0 seconds".

I think we could have also changed the reset to check whether the attempts count was greater than zero and skip reset in that case if we want to avoid returning `nil`.

A larger change we could consider is removing the `rate_limit_period_expired?` check and  `reset_count_and_otp_last_sent_at` call entirely. It's a bit of a relic from using Postgres for the RateLimiter class where the count and expiration would persist far beyond the expiration period. With Redis auto-expiration and the application check, it would be fine to delete. If we did that, we could revert this.